torchao/quantization/subgraph_utils/extract_subgraphs.py

```diff
@@ -0,0 +1,632 @@
+            cur_linear_2_size = weight_shape[1], weight_shape[0]
+
+        printme(f'cur_linear 2 node: {linear_node_2.format_node()}')
+        printme(f'cur_linear 2 mod: {linear_mod_2}')
+        printme(f'cur_linear 2 stack_summary: {get_stack_summary(linear_node_2)}')
+        linear_node_to_use = linear_node_2
+
+    # look at the subsequent ops
+    # note: sometimes this is a view, so might need to look farther
+    printme(f'num users: {len(linear_node_to_use.users)}')
+    for next_n, _ in linear_node_to_use.users.items():
+        for next_n_input in next_n.all_input_nodes:
+            printme(f'next input shape: {get_meta_val(next_n_input).shape}')
+        printme(f'next node: {next_n.format_node()}')
+        printme(f'next stack_summary: {get_stack_summary(next_n)}')
+        if next_n.op == 'call_module':
+            mod = getattr(gm, next_n.target)
+            printme(f'next mod: {mod}')
+            next_node_types.append(type(mod))
+        else:
+            next_node_types.append(next_n.target)
+
+    printme('\ndebugging summary\n')
+    if not dual_linear:
+        linear_shape_str = f'{cur_linear_size}'
+        linear_str = 'Linear'
+    else:
+        linear_shape_str = f'{cur_linear_size} {cur_linear_2_size}'
+        linear_str = 'Linear -> Linear'
+    printme(f'input_shape {prev_input_shape}, (K, N) {linear_shape_str}')
+    subgraph_summary = f'{maybe_short_name(prev_node_type)} -> {linear_str} -> {[maybe_short_name(t) for t in next_node_types]}'
+    printme(subgraph_summary)
+    printme('\n')
+
+    summary_result = [
+        DEBUG_LINEARS_CALL_COUNTER,  # extraction_idx
+        linear_node.target,  # orig_node_name
+        subgraph_idx,
+        cur_linear_size,
+        cur_linear_2_size,
+        subgraph_summary,
+    ]
+    return summary_result
+
+
+def extract_linear_subgraph(
+    old_gm: torch.fx.GraphModule,
+    old_linear_node: torch.fx.Node,
+    old_linear_mod: torch.nn.Module,
+    subgraph_save_filename: str,
+) -> None:
+    """
+    Input: a GraphModule with a `linear_node` calling `linear_mod`.
+
+    This function does the following:
+    * find the subgraph prev_op -> linear_node -> [*next_ops]
+    * create a new GraphModule containing this subgraph
+    * save it to disk for further debugging
+    """
+
+    # to start, create a new module which just calls the linear
+    if old_linear_mod is not None:
+        new_m = torch.nn.Sequential(old_linear_mod)
+    else:
+        weight_val = get_meta_val(old_linear_node.args[1])
+
+        # handle dual linear for inlined here
+        # TODO(future): merge with code below for dual linear for non-inlined
+
+        old_shape = weight_val.shape
+        new_shape = old_shape[1], old_shape[0]
+
+        if not is_first_node_of_dual_linear(old_gm, old_linear_node):
+            new_m = torch.nn.Sequential(
+                torch.nn.Linear(*new_shape, dtype=weight_val.dtype),
+            ).cuda()
+
+        else:
```

Comment:
this file has some complexity for extracting (prev_op -> linear1 -> maybe_linear_2 -> next_ops) because the models we originally studied had back to back linears.  If you only care about transformer models, you can simplify this code quite a bit by removing the special logic for extraction of the second linear.  Happy to point to the right places in the code if needed.

Reply:
let me clean this up a bit later since I'm still not sure if we need to reimplement the functionality with some other approaches yet, will figure out as we expand to test on more models